go/vt/vtgate/plugin_mysql_server.go

```diff
@@ -626,6 +626,9 @@ func newMysqlUnixSocket(address string, authServer mysql.AuthServer, handler mys
 }
 
 func (srv *mysqlServer) shutdownMysqlProtocolAndDrain() {
+	if srv == nil {
+		return
+	}
```

Comment:
Should it be checked higher up the stack? Also avoid future issues if at that level it is used for anything else and it would panic again?

Reply:
@GuptaManan100 I think moving it up is actually right, otherwise we have another panic likely anyway. So something like this:

```go
diff --git i/go/vt/vtgate/vtgate.go w/go/vt/vtgate/vtgate.go
index 4716cdcf79..f3be94599e 100644
--- i/go/vt/vtgate/vtgate.go
+++ w/go/vt/vtgate/vtgate.go
@@ -364,8 +364,10 @@ func Init(
                        st.Start()
                }
                srv := initMySQLProtocol(vtgateInst)
-               servenv.OnTermSync(srv.shutdownMysqlProtocolAndDrain)
-               servenv.OnClose(srv.rollbackAtShutdown)
+               if srv != nil {
+                       servenv.OnTermSync(srv.shutdownMysqlProtocolAndDrain)
+                       servenv.OnClose(srv.rollbackAtShutdown)
+               }
        })
        servenv.OnTerm(func() {
                if st != nil && enableSchemaChangeSignal {
```

Reply:
I am unsure it would provoke a panic because according to the test @GuptaManan100 wrote, the new code handles it well when the `srv` variable is `nil`. @dbussink do you mean it would cause a panic because we are calling a method on a `nil` pointer?
```go
var srv *mysqlServer = nil
srv.shutdownMysqlProtocolAndDrain()
```

Reply:
@frouioui The problem is that this crashes exactly how I described in https://github.com/vitessio/vitess/pull/14902#discussion_r1444224465. The problem is that now this ends up triggering a potential panic:

```
servenv.OnClose(srv.rollbackAtShutdown)
```

In general, `nil` guards inside a struct method are a smell imho. They are sometimes useful, but if they are used, you need to make sure that every function on that struct also does the same `nil` check or otherwise it's way too trivial to introduce again a `panic` on the other functions. That's exactly what's happening here.

Therefore I think it's much safer here to handle this at the caller side, like in my suggested change above. That avoids also the other panic.

Reply:
I have made the change.